qualtran/bloqs/arithmetic/conversions.py

```diff
@@ -150,3 +156,77 @@ def _signed_to_twos() -> SignedIntegerToTwosComplement:
 
 
 _SIGNED_TO_TWOS = BloqDocSpec(bloq_cls=SignedIntegerToTwosComplement, examples=[_signed_to_twos])
+
+
+@frozen
+class ToFxp(Bloq):
+    """Convert a register storing a `QUInt` to a `QFxp`, optionally dividing by a power of two.
+
+    Args:
+        a_dtype: Quantum datatype used to represent the input.
+        b_dtype: Quantum datatype used to represent the output.
+        divide_by: Power of two to divide by when converting (default: 0).
+
+    Registers:
+        a: A a_dtype.bitsize-sized input register (register a above).
+        b: A b_dtype.bitsize-sized output register (register b above).
+    """
+
+    a_dtype: QUInt
+    b_dtype: QFxp = field()
+    divide_by: SymbolicInt = 0
+
+    @b_dtype.validator
+    def _b_dtype_validate(self, field, val):
+        if (
+            is_symbolic(val.bitsize)
+            or is_symbolic(self.a_dtype.bitsize)
+            or is_symbolic(self.divide_by)
+        ):
+            return
+        if val.num_int < self.a_dtype.bitsize - self.divide_by or val.num_frac < self.divide_by:
+            raise ValueError("b_dtype not big enough for a_dtype")
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(a=self.a_dtype, b=self.b_dtype)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {(CNOT(), self.a_dtype.bitsize)}
+
+    def build_composite_bloq(self, bb: BloqBuilder, a: Soquet, b: Soquet) -> Dict[str, SoquetT]:
+        if (
+            is_symbolic(self.a_dtype.bitsize)
+            or is_symbolic(self.b_dtype.num_frac)
+            or is_symbolic(self.divide_by)
+        ):
+            raise DecomposeTypeError(f"Cannot decompose symbolic {self=}.")
+
+        a_bits = bb.split(a)
+        b_bits = bb.split(b)
+        offset = self.b_dtype.num_frac - self.divide_by
+        for i in range(len(a_bits)):
+            a_bits[i], b_bits[offset + i] = bb.add_t(
+                CNOT(), ctrl=a_bits[i], target=b_bits[offset + i]
+            )
+
+        return {"a": bb.join(a_bits), "b": bb.join(b_bits)}
```

Comment:
I think the `ToFxp` bloq should really just be a bookkeeping bloq and we shouldn't do any operations (like CNOT) to implement it. The reason is that to "Convert a register of type QUInt to QFxp" we really  just need to reinterpret where the dot is in the binary representation of the `QUInt`; and this is determined by the `divide_by` parameter. You are not changing the binary string representation of any input (maybe concatenating the input with a bunch of 0's on the left if bitsize of output is greater than input; but I think we should use the `SignExtend` bloq to do the resizing instead of doing it here). 

The current behavior has the unitary $\sum_{a}|a\rangle|b\rangle \rightarrow |a\rangle |b + a\rangle$. Here $a$ is the `QUInt` register and $b$ is the `QFxp` register. Instead what we want is to keep $a$ as it is but just reinterpret it as a `QFxp` instead of a `QUInt`.

Reply:
Cast + Pad maybe?